Polling confirmations (under Experimental spi)

younata opened the pull request:
This implements polling confirmations, as described in [ST-NNNN](https://github.com/younata/swift-evolution/blob/younata/testing-polling-expectations/proposals/testing/NNNN-polling-confirmations.md).

### Motivation:

Being able to monitor changes in the background is something of immense value. Swift Testing already provides an API for this in the `confirmation` api. However, I've found the `confirmation` to be hard to work with at times - it requires you to set up a callback for when something changes, which is not always possible or even the right thing to do. Polling provides a very general approach to monitoring all kinds of changes.

### Modifications:

This adds a new set of macros for handling polling. A new public enum for the 2 separate polling behaviors, and new types to actually implement polling. All under the experimental spi.

### Checklist:

- [x] Code and documentation should follow the style of the [Style Guide](https://github.com/apple/swift-testing/blob/main/Documentation/StyleGuide.md).
- [ ] If public symbols are renamed or modified, DocC references should be updated.

younata committed:
Polling Confirmations

Preliminary implementation of polling expectations

Make Polling into a function along the lines of Confirmation
Additionally, make PollingBehavior an implementation detail of polling, instead of exposed publicly
Removes any timeouts involved for polling, as they become increasingly unreliable
as the system runs more and more tests

Take in configuration arguments, add polling interval

Add traits for configuring polling

Use consistent naming between confirmAlwaysPasses and the related configuration trait
Stop unnecessarily waiting after the last polling attempt has finished.
Allow for subsequent polling configuration traits which specified nil for a value to fall back to earlier polling configuration traits before falling back to the default.

Add requirePassesEventually and requireAlwaysPasses
These two mirror their confirm counterparts, only throwing an error (instead of recording an issue) when they fail.

Rewrite confirmPassesEventually when returning an optional to remove the PollingRecorder actor.
Now, this uses a separate method for evaluating polling to remove that actor.

Clean up the duplicate Poller.evaluate/Poller.poll methods
Removed the duplicate poll method, and made evaluate-returning-bool into a wrapper for evaluate-returning-optional

Configure polling confirmations as timeout & polling interval
This is less direct, but much more intuitive for test authors.
Also add exit tests confirming that these values are non-negative

Rename to actually use the confirmation name
Follow more english-sentence-like guidance for function naming
Simplify the polling confirmation API down to just 2 public functions, 1 enum, and 1 error type.
Always throw an error when polling fails, get rid of the separate issue recording.

Use a single polling confirmation configuration trait
Instead of mulitple traits per stop condition, just have a single trait
per stop condition.

younata commented on Sources/Testing/Polling/Polling.swift:
```diff
@@ -0,0 +1,293 @@
+  ) async -> PollResult {
+    let pollingProcessor = PollingProcessor(behavior: behavior)
+    return await withTaskGroup { taskGroup in
+      taskGroup.addTask {
+        do {
+          try await Task.sleep(for: timeout)
+        } catch {}
+        // Task.sleep will only throw if it's cancelled, at which point this
+        // taskgroup has already returned and we don't care about the value
+        // returned here.
+        return await pollingProcessor.didTimeout()
+      }
```
Per [discussion in forum](https://forums.swift.org/t/pitch-polling-expectations/79866/15), this is unsafe & needs to be rethought.

younata commented on Sources/Testing/Polling/Polling.swift:
```diff
@@ -0,0 +1,293 @@
+
+  /// A type to record the last value returned by a closure returning an optional
+  /// This is only used in the `#require(until:)` macro returning an optional.
+  private actor Recorder<R: Sendable> {
+    var lastValue: R?
+
+    /// Record a new value to be returned
+    func record(value: R) {
+      self.lastValue = value
+    }
+  }
```
Ugh. I really dislike this approach, but I wanted to get this PR up so that others could actually test this proposal out & I couldn't think of a better way to handle this. Hopefully one of the benefits of a non-timed-based approach to polling will be that returning values will be much simpler/not a tacked-on hack.

younata commented on Sources/Testing/Polling/Polling.swift:
```diff
@@ -0,0 +1,293 @@
+    let sourceContext = SourceContext(backtrace: nil, sourceLocation: sourceLocation)
+
+    switch result {
+    case .timedOut:
+      expectation.isPassing = false
+      Issue(
+        kind: .expectationFailed(expectation),
+        comments: comments,
+        sourceContext: sourceContext
+      ).record()
+    case .timedOutWithoutRunning:
+      expectation.isPassing = false
+      Issue(
+        kind: .expectationFailed(expectation),
+        comments: comments,
+        sourceContext: sourceContext
+      ).record()
+    case .finished:
+      return __checkValue(
+        true,
+        expression: expression,
+        comments: comments,
+        isRequired: isRequired,
+        sourceLocation: sourceLocation
+      )
+    case .failed:
+      return __checkValue(
+        false,
+        expression: expression,
+        comments: comments,
+        isRequired: isRequired,
+        sourceLocation: sourceLocation
+      )
+    case .cancelled:
+      Issue(
+        kind: .system,
+        comments: comments,
+        sourceContext: sourceContext
+      ).record()
+    }
```
a) I'm going to deduplicate this analysis with the nearly-identical section in the other version of `run`. But time constraints & a desire to let others try out this code were why this was left in.
b) I'd really like to be able to provide additional explanations behind why each issue is recorded, but it wasn't clear what additional arguments to Issue (if any, maybe I need to set a property and I missed that?) I should use to provide that additional context.

younata commented on Tests/TestingTests/PollingTests.swift:
```diff
@@ -0,0 +1,352 @@
+    }
+
+    @Test("Simple failing expressions") func trivialSadPath() async {
+      await confirmation("Polling failed", expectedCount: 1) { failed in
+        var configuration = Configuration()
+        configuration.eventHandler = { event, _ in
+          if case .issueRecorded = event.kind {
+            failed()
+          }
+        }
+        await Test {
+          await #expect(until: .passesOnce) { false }
+        }.run(configuration: configuration)
+      }
```
I'm going to write a test helper to make it much easier to verify that issues were reported, as `withKnownIssue` is the wrong semantics for what I want. I'm also tempted to make it available under `ForToolsIntegrationOnly`, but I also think that should be a separate pitch?

I really want something that I can use like:

```swift
let issues = await collectIssues {
    #expect(Bool(false))
}
```

younata commented on Tests/TestingTests/PollingTests.swift:
```diff
@@ -0,0 +1,352 @@
+}
+
+extension DurationProtocol {
+  fileprivate func isCloseTo(other: Self, within delta: Self) -> Bool {
+    var distance = self - other
+    if (distance < Self.zero) {
+      distance *= -1
+    }
+    return distance <= delta
+  }
```
I want this to be in the swift standard library and apply to all numeric types. It would make comparing floating point types so much more doable.

In my ideal world, this would work like `1.000001 == 1 ± 1e-3` (or `1.00001 == 1 +- 1e-3`, to avoid using the hard-to-type ± symbol). But that would involve adding another ternary operator to the language, and would be much more involved that adding a method to a protocol in the standard library.

grynspan replied:
swift-numerics has this, more or less, but we can't add it as a dependency of course. :)

younata commented on Sources/Testing/Issues/Issue.swift:
```diff
@@ -38,6 +38,15 @@ public struct Issue: Sendable {
+    /// ``confirmation(_:until:within:pollingEvery:isolation:sourceLocation:_:)-5tnlk``
+    /// whenever the polling fails, as described in ``PollingStopCondition``.
+    @_spi(Experimental)
+    case pollingConfirmationFailed
```
That seems reasonable to do. Maybe for if an error was thrown, and if the stopping condition fails?

younata committed:
Include a failure reason to help the console figure out why a polling confirmation failed

plemarquand reviewed (commented):
Overall looking really great. Just one comment and a bunch of documentation nits.

younata committed:
Polling confirmations: Handle extremely large polling iterations

younata committed:
Polling Confirmations: Documentation and spelling mistakes

younata committed:
Polling confirmations: Pass the license headers check

stmontgomery commented on Sources/Testing/Polling/Polling.swift:
```diff
@@ -0,0 +1,463 @@
+
+/// Default values for polling confirmations.
+@available(_clockAPI, *)
+internal let defaultPollingConfiguration = (
```
Can this be `private`?

```suggestion
private let _defaultPollingConfiguration = (
```

stmontgomery commented on Sources/Testing/Polling/Polling.swift:
New files, such as this one and `PollingConfigurationTrait.swift`, need to be added to the CMake scripts. Can you be sure to add them there? See the file list in `Sources/Testing/CMakeLists.txt`.

stmontgomery commented on Sources/Testing/Polling/Polling.swift:
In terms of the placement of this new file, I'd like to suggest instead of creating a new `Polling/` subdirectory we create a new `Confirmations/` subdirectory, add move the existing `Confirmation.swift` to that new directory, and place this new file there too.

(And if you _do_ take this suggestion and move some files around, be sure to adjust the CMake files accordingly.)

stmontgomery commented on Sources/Testing/Polling/Polling.swift:
```diff
@@ -0,0 +1,463 @@
+
+/// A type describing why polling failed
+@_spi(Experimental)
+public enum PollingFailureReason: Sendable, Codable {
```
I think this `PollingFailureReason` enum would feel more natural nested within `PollingFailedError`, and then its name could be simplified to `Reason` (or fully-qualified, `PollingFailedError.Reason`).

younata replied:
Fair idea. The main reason for keeping `PollingFailedError` a struct is to make the conformance to `CustomIssueRepresentable` simple and obvious.

For what test authors consume, I expect `PollingFailedError.Reason` provides enough information for them to do anything custom - they can't really access `sourceContext` in the general case, and `comment` likely isn't that useful to handle?

I'm inclined to keep with @stmontgomery's suggestion right now, but I definitely see @grynspan's argument.

younata committed:
PR Feedback

Mark defaultPollingConfiguration as private
Add polling files to CMakeLists
Move Polling and Confirmations.swift to a new Confirmations directory
Rename PollingFailureReason to PallingFailedError.Reason

jerryjrchen commented on Sources/Testing/Confirmations/Polling.swift:
```diff
@@ -0,0 +1,462 @@
+///     with a matching stopCondition.
+///     If no such trait has been added, then polling will be attempted for
+///     about 1 second before recording an issue.
+///     `duration` must be greater than 0.
```
Might also be worth mentioning that `duration > interval` based on the current implementation.

Although from a user perspective, should `duration == interval` be permissible? I would interpret that basically a single poll (so perhaps not very useful), but still a possible use case.

younata replied:
Thanks for bringing this up!

I'll update the documentation & add some tests mentioning this.

I don't see any real harm in disallowing the case where `duration == interval`. It is not very useful - at least not at the moment - but to your point, it is a possible use case someone might have.

jerryjrchen commented:
The link in the description links to the old path for the proposal doc. I think it should be updated to point to https://github.com/younata/swift-evolution/blob/younata/testing-polling-expectations/proposals/testing/NNNN-polling-confirmations.md instead

(I think I have edit permissions to make this description change but wasn't sure if that would be polite to do 😅 )

jerryjrchen commented on Sources/Testing/Confirmations/Polling.swift:
```diff
@@ -0,0 +1,462 @@
+/// - Parameters:
+///   - comment: A user-specified comment describing this confirmation.
+///   - stopCondition: When to stop polling.
+///   - duration: The expected length of time to continue polling for.
```
Would it also be worth mentioning that time to run the body isn't accounted for in `duration`? I can see someone unfamiliar with the full details of polling confirmations writing something like:
```swift
      try await confirmation(until: .firstPass, within: .seconds(10)) {
        tenSecondLongOperation()
      }
```

And then being surprised it runs much longer than 10 seconds total.

jerryjrchen commented on Tests/TestingTests/PollingTests.swift:
Thanks for the awesome comprehensive tests!

jerryjrchen commented on Tests/TestingTests/PollingTests.swift:
```diff
@@ -0,0 +1,556 @@
+          await incrementor.increment() != 0
+        }
+        let count = await incrementor.count
+        #expect(await count == 10)
```
I think this `await` is unnecessary:
```suggestion
        #expect(count == 10)
```

jerryjrchen commented on Tests/TestingTests/PollingTests.swift:
```diff
@@ -0,0 +1,556 @@
+        #expect(await incrementor.count == 50)
+      }
+
+#if !SWT_NO_EXIT_TESTS
```
Nice use of exit tests!

jerryjrchen reviewed (approved):
Overall this looks great to me, and I mostly had a bunch of nitpicks about the docs. I'll caveat by saying I've only read through the proposal recently and skimmed the pitch forum post, so I hope I didn't rehash a bunch of previous questions :)

younata commented on Sources/Testing/Confirmations/Polling.swift:
```diff
@@ -0,0 +1,462 @@
+      if let result {
+        return .succeeded(result)
+      } else {
+        return .continuePolling
```
I appreciate the feedback, and while pedantic, I do agree that makes it easier to read.

I think this is a good suggestion for clarity purposes. It does make the last line in `poll` redundant (`return .failed` after the for loop exits), but necessary to satisfy the compiler. I'll add a comment to the end of that method explaining this.

younata committed:
PR and documentation feedback

younata committed:
Update documentation for PollingConfigurationTrait to reflect that duration may be greater than or equal to interval

younata committed:
Polling confirmations: Remove the isolation paramater

Have the polling bodies themselves specify nonisolated(nonsending).
Keep the escaping attribute for polling bodies. While not strictly necessary, I strongly suspect that a future refactor to use task execution time will make that requirement necessary.

younata committed:
Polling confirmations: Add ability to include comments about why a particular polling attempt failed.

younata committed:
Polling confirmations: Add ability to include comments about why a particular polling attempt failed.

younata committed:
Introduce workaround for xcode not indicating a failure when new Issue.Kind.Snapshot cases are used

For now, polling confirmation failures will be reported in the snapshot as unconditional failures.
